### PR TITLE
ocamlPackages.cstruct: 5.0.0 -> 6.0.0

### PR DESCRIPTION
**Pull request by sternenseemann**
###### Motivation for this change

<https://github.com/mirage/ocaml-cstruct/releases/tag/v6.0.0>, switch to ppxlib.

This PR also contains commits from #108137 (which made this upgrade possible) and should probably reviewed after the mentioned PR has been merged. For now this is for informational purposes mostly.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sternenseemann** [comment]
Result of `nixpkgs-review pr 108492` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>ocamlPackages.irmin-http</li>
    <li>ocamlPackages.irmin-unix</li>
  </ul>
</details>
<details>
  <summary>136 packages built:</summary>
  <ul>
    <li>beluga</li>
    <li>coqPackages.coq-elpi</li>
    <li>coqPackages.hierarchy-builder</li>
    <li>dune-release</li>
    <li>flow</li>
    <li>fstar</li>
    <li>jackline</li>
    <li>libbap</li>
    <li>liquidsoap</li>
    <li>ocamlPackages.arp</li>
    <li>ocamlPackages.asn1-combinators</li>
    <li>ocamlPackages.bisect_ppx</li>
    <li>ocamlPackages.bitstring</li>
    <li>ocamlPackages.ca-certs</li>
    <li>ocamlPackages.cohttp-async</li>
    <li>ocamlPackages.cohttp-lwt-unix</li>
    <li>ocamlPackages.conduit</li>
    <li>ocamlPackages.conduit-async</li>
    <li>ocamlPackages.conduit-lwt</li>
    <li>ocamlPackages.conduit-lwt-unix</li>
    <li>ocamlPackages.cow</li>
    <li>ocamlPackages.cstruct</li>
    <li>ocamlPackages.cstruct-lwt</li>
    <li>ocamlPackages.cstruct-sexp</li>
    <li>ocamlPackages.cstruct-unix</li>
    <li>ocamlPackages.curly</li>
    <li>ocamlPackages.digestif</li>
    <li>ocamlPackages.dns</li>
    <li>ocamlPackages.dns-certify</li>
    <li>ocamlPackages.dns-cli</li>
    <li>ocamlPackages.dns-client</li>
    <li>ocamlPackages.dns-mirage</li>
    <li>ocamlPackages.dns-resolver</li>
    <li>ocamlPackages.dns-server</li>
    <li>ocamlPackages.dns-stub</li>
    <li>ocamlPackages.dns-tsig</li>
    <li>ocamlPackages.duff</li>
    <li>ocamlPackages.eliom</li>
    <li>ocamlPackages.elpi</li>
    <li>ocamlPackages.eqaf</li>
    <li>ocamlPackages.ethernet</li>
    <li>ocamlPackages.ezjsonm</li>
    <li>ocamlPackages.fiat-p256</li>
    <li>ocamlPackages.git</li>
    <li>ocamlPackages.git-http</li>
    <li>ocamlPackages.git-unix</li>
    <li>ocamlPackages.graphql-cohttp</li>
    <li>ocamlPackages.graphql_ppx</li>
    <li>ocamlPackages.hacl_x25519</li>
    <li>ocamlPackages.hex</li>
    <li>ocamlPackages.hkdf</li>
    <li>ocamlPackages.io-page</li>
    <li>ocamlPackages.io-page-unix</li>
    <li>ocamlPackages.ipaddr-cstruct</li>
    <li>ocamlPackages.ipaddr-sexp</li>
    <li>ocamlPackages.iri</li>
    <li>ocamlPackages.irmin</li>
    <li>ocamlPackages.irmin-chunk</li>
    <li>ocamlPackages.irmin-fs</li>
    <li>ocamlPackages.irmin-git</li>
    <li>ocamlPackages.irmin-graphql</li>
    <li>ocamlPackages.irmin-mem</li>
    <li>ocamlPackages.irmin-pack</li>
    <li>ocamlPackages.irmin-test</li>
    <li>ocamlPackages.jingoo</li>
    <li>ocamlPackages.jwto</li>
    <li>ocamlPackages.lwt_ppx</li>
    <li>ocamlPackages.macaddr-cstruct</li>
    <li>ocamlPackages.macaddr-sexp</li>
    <li>ocamlPackages.metrics-mirage</li>
    <li>ocamlPackages.mirage-block</li>
    <li>ocamlPackages.mirage-block-combinators</li>
    <li>ocamlPackages.mirage-block-ramdisk</li>
    <li>ocamlPackages.mirage-block-unix</li>
    <li>ocamlPackages.mirage-channel</li>
    <li>ocamlPackages.mirage-console</li>
    <li>ocamlPackages.mirage-console-unix</li>
    <li>ocamlPackages.mirage-crypto</li>
    <li>ocamlPackages.mirage-crypto-pk</li>
    <li>ocamlPackages.mirage-crypto-rng</li>
    <li>ocamlPackages.mirage-crypto-rng-async</li>
    <li>ocamlPackages.mirage-crypto-rng-mirage</li>
    <li>ocamlPackages.mirage-flow</li>
    <li>ocamlPackages.mirage-flow-combinators</li>
    <li>ocamlPackages.mirage-flow-unix</li>
    <li>ocamlPackages.mirage-fs</li>
    <li>ocamlPackages.mirage-logs</li>
    <li>ocamlPackages.mirage-nat</li>
    <li>ocamlPackages.mirage-net</li>
    <li>ocamlPackages.mirage-profile</li>
    <li>ocamlPackages.mirage-protocols</li>
    <li>ocamlPackages.mirage-random</li>
    <li>ocamlPackages.mirage-random-test</li>
    <li>ocamlPackages.mirage-stack</li>
    <li>ocamlPackages.mirage-types</li>
    <li>ocamlPackages.mirage-types-lwt</li>
    <li>ocamlPackages.mirage-unix</li>
    <li>ocamlPackages.mirage-vnetif</li>
    <li>ocamlPackages.mustache</li>
    <li>ocamlPackages.nocrypto</li>
    <li>ocamlPackages.ocaml-monadic</li>
    <li>ocamlPackages.ocsigen-start</li>
    <li>ocamlPackages.ocsigen-toolkit</li>
    <li>ocamlPackages.ocsigen_server</li>
    <li>ocamlPackages.opium</li>
    <li>ocamlPackages.opium_kernel</li>
    <li>ocamlPackages.otr</li>
    <li>ocamlPackages.pcap-format</li>
    <li>ocamlPackages.pgocaml</li>
    <li>ocamlPackages.pgocaml_ppx</li>
    <li>ocamlPackages.phylogenetics</li>
    <li>ocamlPackages.piqi</li>
    <li>ocamlPackages.piqi-ocaml</li>
    <li>ocamlPackages.ppx_cstruct</li>
    <li>ocamlPackages.ppx_deriving</li>
    <li>ocamlPackages.ppx_deriving-4-5</li>
    <li>ocamlPackages.ppx_deriving_rpc</li>
    <li>ocamlPackages.ppx_deriving_yojson</li>
    <li>ocamlPackages.ppx_import</li>
    <li>ocamlPackages.ppx_tools_versioned</li>
    <li>ocamlPackages.prof_spacetime</li>
    <li>ocamlPackages.randomconv</li>
    <li>ocamlPackages.reason</li>
    <li>ocamlPackages.rpclib-lwt</li>
    <li>ocamlPackages.sedlex</li>
    <li>ocamlPackages.sedlex_2</li>
    <li>ocamlPackages.tcpip</li>
    <li>ocamlPackages.tls</li>
    <li>ocamlPackages.visitors</li>
    <li>ocamlPackages.x509</li>
    <li>ocamlPackages.xtmpl</li>
    <li>python37Packages.bap</li>
    <li>python38Packages.bap</li>
    <li>python39Packages.bap</li>
    <li>satysfi</li>
    <li>stog</li>
  </ul>
</details>

File: pkgs/development/ocaml-modules/ptime/default.nix [hunk 1]
```diff
@@ -10,11 +12,12 @@ stdenv.mkDerivation rec {
   };
 
   nativeBuildInputs = [ ocaml findlib ocamlbuild ];
-  buildInputs = [ findlib topkg js_of_ocaml ];
+  buildInputs = [ findlib topkg ]
+    ++ stdenv.lib.optional jsooSupport js_of_ocaml;
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/ptime/default.nix:16]:
```suggestion
    ++ lib.optional jsooSupport js_of_ocaml;
```

File: pkgs/development/ocaml-modules/ptime/default.nix [hunk 2]
```diff
@@ -10,11 +12,12 @@ stdenv.mkDerivation rec {
...
 
   propagatedBuildInputs = [ result ];
 
-  buildPhase = "${topkg.run} build --with-js_of_ocaml true";
+  buildPhase = "${topkg.run} build --with-js_of_ocaml ${stdenv.lib.boolToString jsooSupport}";
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/ptime/default.nix:20]:
```suggestion
  buildPhase = "${topkg.run} build --with-js_of_ocaml ${lib.boolToString jsooSupport}";
```

Comment (sternenseemann) [reply pkgs/development/ocaml-modules/ptime/default.nix]: Thanks!

**Review by vbgl — CHANGES_REQUESTED** [review]
`xenstore` should now propagate `stdlib-shims`

**Comment by sternenseemann** [comment]
> xenstore should now propagate stdlib-shims

#111507

**Comment by vbgl** [comment]
Why is this in a separate PR? Isn’t it related to the update of cstruct/ppx_cstruct?

**Comment by sternenseemann** [comment]
Yeah, just didn't realize it was related! Added it here as well.

**Comment by sternenseemann** [comment]
Result of `nixpkgs-review pr 108492` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>121 packages built:</summary>
  <ul>
    <li>dune-release</li>
    <li>jackline</li>
    <li>libbap</li>
    <li>ocamlPackages.arp</li>
    <li>ocamlPackages.asn1-combinators</li>
    <li>ocamlPackages.awa</li>
    <li>ocamlPackages.awa-lwt</li>
    <li>ocamlPackages.awa-mirage</li>
    <li>ocamlPackages.ca-certs</li>
    <li>ocamlPackages.cohttp-async</li>
    <li>ocamlPackages.cohttp-lwt-unix</li>
    <li>ocamlPackages.cohttp-mirage</li>
    <li>ocamlPackages.conduit</li>
    <li>ocamlPackages.conduit-async</li>
    <li>ocamlPackages.conduit-lwt</li>
    <li>ocamlPackages.conduit-lwt-unix</li>
    <li>ocamlPackages.conduit-mirage</li>
    <li>ocamlPackages.cow</li>
    <li>ocamlPackages.cstruct</li>
    <li>ocamlPackages.cstruct-lwt</li>
    <li>ocamlPackages.cstruct-sexp</li>
    <li>ocamlPackages.cstruct-unix</li>
    <li>ocamlPackages.curly</li>
    <li>ocamlPackages.digestif</li>
    <li>ocamlPackages.dns</li>
    <li>ocamlPackages.dns-certify</li>
    <li>ocamlPackages.dns-cli</li>
    <li>ocamlPackages.dns-client</li>
    <li>ocamlPackages.dns-mirage</li>
    <li>ocamlPackages.dns-resolver</li>
    <li>ocamlPackages.dns-server</li>
    <li>ocamlPackages.dns-stub</li>
    <li>ocamlPackages.dns-tsig</li>
    <li>ocamlPackages.duff</li>
    <li>ocamlPackages.eliom</li>
    <li>ocamlPackages.eqaf</li>
    <li>ocamlPackages.ethernet</li>
    <li>ocamlPackages.ezjsonm</li>
    <li>ocamlPackages.fiat-p256</li>
    <li>ocamlPackages.git</li>
    <li>ocamlPackages.git-http</li>
    <li>ocamlPackages.git-unix</li>
    <li>ocamlPackages.graphql-cohttp</li>
    <li>ocamlPackages.hacl_x25519</li>
    <li>ocamlPackages.hex</li>
    <li>ocamlPackages.hkdf</li>
    <li>ocamlPackages.io-page</li>
    <li>ocamlPackages.io-page-unix</li>
    <li>ocamlPackages.ipaddr-cstruct</li>
    <li>ocamlPackages.ipaddr-sexp</li>
    <li>ocamlPackages.irmin</li>
    <li>ocamlPackages.irmin-chunk</li>
    <li>ocamlPackages.irmin-fs</li>
    <li>ocamlPackages.irmin-git</li>
    <li>ocamlPackages.irmin-graphql</li>
    <li>ocamlPackages.irmin-http</li>
    <li>ocamlPackages.irmin-mem</li>
    <li>ocamlPackages.irmin-pack</li>
    <li>ocamlPackages.irmin-test</li>
    <li>ocamlPackages.irmin-unix</li>
    <li>ocamlPackages.macaddr-cstruct</li>
    <li>ocamlPackages.macaddr-sexp</li>
    <li>ocamlPackages.metrics-mirage</li>
    <li>ocamlPackages.mimic</li>
    <li>ocamlPackages.mirage-block</li>
    <li>ocamlPackages.mirage-block-combinators</li>
    <li>ocamlPackages.mirage-block-ramdisk</li>
    <li>ocamlPackages.mirage-block-unix</li>
    <li>ocamlPackages.mirage-channel</li>
    <li>ocamlPackages.mirage-console</li>
    <li>ocamlPackages.mirage-console-unix</li>
    <li>ocamlPackages.mirage-crypto</li>
    <li>ocamlPackages.mirage-crypto-pk</li>
    <li>ocamlPackages.mirage-crypto-rng</li>
    <li>ocamlPackages.mirage-crypto-rng-async</li>
    <li>ocamlPackages.mirage-crypto-rng-mirage</li>
    <li>ocamlPackages.mirage-flow</li>
    <li>ocamlPackages.mirage-flow-combinators</li>
    <li>ocamlPackages.mirage-flow-unix</li>
    <li>ocamlPackages.mirage-fs</li>
    <li>ocamlPackages.mirage-logs</li>
    <li>ocamlPackages.mirage-nat</li>
    <li>ocamlPackages.mirage-net</li>
    <li>ocamlPackages.mirage-profile</li>
    <li>ocamlPackages.mirage-protocols</li>
    <li>ocamlPackages.mirage-random</li>
    <li>ocamlPackages.mirage-random-test</li>
    <li>ocamlPackages.mirage-stack</li>
    <li>ocamlPackages.mirage-types</li>
    <li>ocamlPackages.mirage-types-lwt</li>
    <li>ocamlPackages.mirage-unix</li>
    <li>ocamlPackages.mirage-vnetif</li>
    <li>ocamlPackages.mustache</li>
    <li>ocamlPackages.nocrypto</li>
    <li>ocamlPackages.ocsigen-start</li>
    <li>ocamlPackages.ocsigen-toolkit</li>
    <li>ocamlPackages.ocsigen_server</li>
    <li>ocamlPackages.opium</li>
    <li>ocamlPackages.opium_kernel</li>
    <li>ocamlPackages.otr</li>
    <li>ocamlPackages.pcap-format</li>
    <li>ocamlPackages.pgocaml</li>
    <li>ocamlPackages.pgocaml_ppx</li>
    <li>ocamlPackages.ppx_cstruct</li>
    <li>ocamlPackages.ppx_repr</li>
    <li>ocamlPackages.prof_spacetime</li>
    <li>ocamlPackages.randomconv</li>
    <li>ocamlPackages.tcpip</li>
    <li>ocamlPackages.tls</li>
    <li>ocamlPackages.tls-mirage</li>
    <li>ocamlPackages.vchan</li>
    <li>ocamlPackages.wodan</li>
    <li>ocamlPackages.wodan-irmin</li>
    <li>ocamlPackages.wodan-unix</li>
    <li>ocamlPackages.x509</li>
    <li>ocamlPackages.xenstore</li>
    <li>ocamlPackages.xenstore-tool</li>
    <li>ocamlPackages.xenstore_transport</li>
    <li>python37Packages.bap</li>
    <li>python38Packages.bap</li>
    <li>python39Packages.bap</li>
  </ul>
</details>

**Review by vbgl — APPROVED** [review]

**Pull request merged**